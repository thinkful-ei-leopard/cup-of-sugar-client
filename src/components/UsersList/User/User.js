import React from 'react';
import { withRouter } from 'react-router-dom';
import ThreadsContext from '../../../contexts/ThreadsContext';
import ThreadsApiService from '../../../services/threads-api-service';
import styles from './User.module.scss';

class User extends React.Component {
  static contextType = ThreadsContext;

  redirectToThread = threadId => {
    this.props.history.push(`/thread/${threadId}`);
  };

  async handleThreadCreate() {
    const user = this.props.user;
    const neighbor = this.props.neighbor;
    let threads = this.context.threads;
    let thread = threads.find(
      thread =>
        (thread.user_id1 === user.id || thread.user_id2 === user.id) &&
        (thread.user_id1 === neighbor.id || thread.user_id2 === neighbor.id)
    );
    if (thread) {
      this.redirectToThread(thread.id);
      return;
    }
    let newThread = {
      user_id2: neighbor.id,
      name1: user.name,
      user_name1: user.username,
      name2: neighbor.name,
      user_name2: neighbor.user_name,
      img_src1: user.img_src,
      img_alt1: user.img_alt,
      img_src2: neighbor.img_src || require('../../../images/user.png'),
      img_alt2: neighbor.img_alt
    };
    thread = await ThreadsApiService.addThread(newThread);
    this.redirectToThread(thread.id);
  }

  render() {
    const { neighbor } = this.props;
    return (
      <li
        title="Start a Thread"
        className={styles.userListItem}
        onClick={e => {
          e.preventDefault();
          this.handleThreadCreate();
        }}>
        <p className={styles.neighbor}>
          {neighbor.name} ({neighbor.user_name})
        </p>
      </li>
    );
  }
}

export default withRouter(User);
