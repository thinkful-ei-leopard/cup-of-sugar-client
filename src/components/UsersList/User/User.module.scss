@import '../../../styles/_breakpoints.scss';
@import '../../../styles/_variables.scss';

.userListItem {
  list-style-type: none;
  padding: 0.5rem 2rem;
  border-radius: 1rem;
  margin: 0 auto;
}

.neighbor {
  font-size: 1.2rem;
  color: var(--dark2);
  transition: color 0.2s;
  &:hover {
    color: var(--dark1);
    cursor: pointer;
  }
}
