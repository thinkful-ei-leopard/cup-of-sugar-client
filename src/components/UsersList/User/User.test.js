import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import User from './User';

it('renders without crashing', () => {
  const div = document.createElement('div');

  ReactDOM.render(
    <Router>
      <User
        history={{}}
        location={{}}
        match={{}}
        neighbor={{}}
        staticContext={undefined}
        user={{}}
      />
    </Router>,
    div
  );

  ReactDOM.unmountComponentAtNode(div);
});

it('renders the UI as expected', () => {
  const tree = renderer
    .create(
      <Router>
        <User
          history={{}}
          location={{}}
          match={{}}
          neighbor={{}}
          staticContext={undefined}
          user={{}}
        />
      </Router>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
