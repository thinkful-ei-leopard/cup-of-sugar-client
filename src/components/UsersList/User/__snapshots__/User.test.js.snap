// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the UI as expected 1`] = `
<li
  className="userListItem"
  onClick={[Function]}
  title="Start a Thread"
>
  <p
    className="neighbor"
  >
    <img
      className="neighborDirectoryAvatar"
    />
     (
    )
  </p>
</li>
`;
